src/renderer/renderScene.ts

```diff
@@ -73,6 +73,31 @@ const strokeCircle = (
   context.stroke();
 };
 
+const renderGrid = (
+  context: CanvasRenderingContext2D,
+  gridSize: number,
+  offsetX: number,
+  offsetY: number,
+  width: number,
+  height: number,
+) => {
+  const origStrokeStyle = context.strokeStyle;
+  context.strokeStyle = "rgba(0,0,0,0.1)";
+  for (let x = offsetX; x < offsetX + width + gridSize; x += gridSize) {
+    context.beginPath();
+    context.moveTo(x, offsetY - gridSize);
+    context.lineTo(x, offsetY + height + gridSize);
+    context.stroke();
+  }
+  for (let y = offsetY; y < offsetY + height + gridSize; y += gridSize) {
+    context.beginPath();
+    context.moveTo(offsetX - gridSize, y);
+    context.lineTo(offsetX + width + gridSize, y);
+    context.stroke();
+  }
+  context.strokeStyle = origStrokeStyle;
+};
```

Comment:
you can increase the grid render performance by doing a single `stroke`.  Fewer drawing operation the better, every time. (on large screen goes from 0.25ms to 0.05ms) May not be noticeable but every ns counts when drawing :)

```javascript
context.beginPath();
for (let x = offsetX; x < offsetX + width + gridSize; x += gridSize) {
  context.moveTo(x, offsetY - gridSize);
  context.lineTo(x, offsetY + height + gridSize);
}
for (let y = offsetY; y < offsetY + height + gridSize; y += gridSize) {
  context.moveTo(offsetX - gridSize, y);
  context.lineTo(offsetX + width + gridSize, y);
}
context.stroke();
```

Btw, since grid never changes per se, we should consider some way of not having to render it every time.

Reply:
Thanks @pshihn ! Fixed in f4da892.

> Btw, since grid never changes per se, we should consider some way of not having to render it every time.

That's not only for grid, but all shapes.

Reply:
Maybe I can revisit https://github.com/excalidraw/excalidraw/pull/780 which I closed because the multiplayer didn't play well with that heuristic. Maybe it can be worked around, dunno.

Reply:
Another crazy idea is to use webgl, which I've experimented in [excalidraw-layers](https://github.com/dai-shi/excalidraw-layers). I'm not suggesting it for the current excalidraw, but it might be interesting in general to explore utilizing webgl for 2D.